music21/voiceLeading.py

```diff
@@ -518,7 +518,9 @@ def parallelMotion(
         >>> vl.parallelMotion(gi, allowOctaveDisplacement=True)
         True
         '''
-
+        # .generic could only be None if intervals were constructed without diatonic components
+        assert self.vIntervals[0].generic is not None
+        assert self.vIntervals[1].generic is not None
```

Comment:
Let's find a way to work around this.  I **hate** bare asserts.

Reply:
It's all good. Not a problem to get rid of em.

In this case, my adding an assert here was just lazy, because there _is_ a possible path for a user to get here. They could create voice leading quartets made up of totally empty intervals. So mypy found a corner case bug, and I was too lazy to add a proper exception. As opposed to the one below about `common.isIterable` where the so-called failure case is not reachable because of what we know about `isIterable()`, that a boolean should never pass that function (it's music21's very own type-narrowing function).

To think aloud just to check my own understanding. (Not to persuade, because let's definitely not add things that you find annoying!)

The `assert` doesn't bother me as an articulation of an assumption the author relies on down the line. ("Somewhere else in the module is responsible for making this not None. But that's forty lines away, and if it ever gets refactored, let's add safety here, because I'm about to start accessing attributes on it.") As opposed to a failure path that is part of normal execution that a user might actually need to know about, care about, and catch. And I think part and parcel with `no-implication-that-users-should-know-or-care` is why you can run python in `-O` to skip them. They're not necessarily even net negatives for user feedback, if the author is proven wrong and the "never supposed to happen" turns out to happen--as it's better to get `AssertionError: self.stream is None`, which articulates the root cause, instead of `AttributeError: self.stream has no attribute __iter__` 400 lines later, which obscures it. Yes, more messages for the user is always nice, but this is the tool for those gray areas "no, this can't fail, but it wouldn't be safe later if something else changes". If there is a path for a user to reach it, then assert is not the tool.

In a few weeks I can spruce these up to disable/raise instead.

I think the value of `mypy` is not just "did it find bugs in my functionality" but maybe more importantly for us "did it find bugs in my type annotations".  On a project with community contributions, it should save a future maintainer time to make sure folks aren't adding in new bad annotations.

Reply:
None of this is new to you, certainly! Figuring out what is safe to rely on versus what needs to be guarded is an everyday part of programming. I guess one to look at it when developing a library is that by adding explicit guards they can sort of start to look like part of the API. Function X raises exception Y. When truly that's not the case because there should be no possible way to get there. And so an assert more clearly shows that it's not part of the API; doesn't represent a test case that needs to be supported after a refactor. Something like that.

Reply:
ah, what a funny memory, ten minutes ago, that time I free-associated about "everyday programming" to someone with 10x my experience... 😉

Reply:
Will reply more soon. But my longer experience is more than exceeded by your talent and enthusiasm.

Reply:
(Thanks.)

Raising exceptions didn't work here either without a lot of boilerplate (you apparently have to narrow in the same scope), so will just disable these ones.